### PR TITLE
[SPARK-5598][MLLIB] model save/load for ALS

### PR DESCRIPTION
**Pull request by mengxr**
following #4233. @jkbradley

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala [hunk 1]
```diff
@@ -136,3 +147,69 @@ class MatrixFactorizationModel(
...
+  import org.apache.spark.mllib.util.Loader._
+
+  override def load(sc: SparkContext, path: String): MatrixFactorizationModel = {
+    val (loadedClassName, formatVersion, metadata) = loadMetadata(sc, path)
+    val classNameV1_0 = SaveLoadV1_0.thisClassName
+    (loadedClassName, formatVersion) match {
+      case (className, "1.0") if className == classNameV1_0 =>
+        SaveLoadV1_0.load(sc, path)
+      case _ =>
+        throw new IOException("" +
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: Is this the preferred syntax?  I've been wondering about this

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala [hunk 2]
```diff
@@ -136,3 +147,69 @@ class MatrixFactorizationModel(
     scored.top(num)(Ordering.by(_._2))
   }
 }
+
+private object MatrixFactorizationModel extends Loader[MatrixFactorizationModel] {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: not private

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala [hunk 3]
```diff
@@ -125,6 +130,12 @@ class MatrixFactorizationModel(
     recommend(productFeatures.lookup(product).head, userFeatures, num)
       .map(t => Rating(t._1, product, t._2))
 
+  override val formatVersion: String = "1.0"
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: protected

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala [hunk 4]
```diff
@@ -136,3 +147,69 @@ class MatrixFactorizationModel(
...
+        SaveLoadV1_0.load(sc, path)
+      case _ =>
+        throw new IOException("" +
+          "MatrixFactorizationModel.load did not recognize model with" +
+          s"(class: $loadedClassName, version: $formatVersion). Supported:\n" +
+          s"  ($classNameV1_0, 1.0)")
+    }
+  }
+
+  private object SaveLoadV1_0 extends Loader[MatrixFactorizationModel] {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: Does not need to extend Loader

**Comment by jkbradley** [comment]
I just had a few small comments.  One question which occurred to me: Does SQLContext have configuration info which would be relevant to saving/loading?  I'm wondering if users should be able to pass in a SQLContext to save/load.

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala (hunk 1 above)
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: Here can't you just omit the `"" +`? maybe it was just auto-inserted by the IDE on hitting return there.

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: I assume it's to make the lines below line up

Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: Yes, that's was from IDE.

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala (hunk 2 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: nice catch

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala (hunk 3 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: done

File: mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala (hunk 4 above)
Comment (mengxr) [inline mllib/src/main/scala/org/apache/spark/mllib/recommendation/MatrixFactorizationModel.scala]: This should be okay. It is indeed a Loader, so I can omit the doc of `load` ....

**Comment by mengxr** [comment]
This is the `spark.mllib` package. We don't expect users to use Spark SQL here.

**Comment by jkbradley** [comment]
Having SaveLoadV1_0 be a Loader is weird since it loads metadata, and the metadata gives the version number.  It works now since there is only 1 version, but it will have to stop being a Loader in the future.

**Comment by jkbradley** [comment]
Other than the Loader issue, looks good.

**Comment by mengxr** [comment]
Merged into master and branch-1.3.

**Pull request closed without merging**